src/Subscription.hpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * This software is distributed under the terms of the MIT License.
+ * Copyright (c) 2020 LXRobotics.
+ * Author: Alexander Entinger <alexander.entinger@lxrobotics.com>
+ * Contributors: https://github.com/107-systems/107-Arduino-Cyphal/graphs/contributors.
+ */
+
+#ifndef INC_107_ARDUINO_CYPHAL_SUBSCRIPTION_HPP
+#define INC_107_ARDUINO_CYPHAL_SUBSCRIPTION_HPP
+
+/**************************************************************************************
+ * INCLUDE
+ **************************************************************************************/
+
+#include <memory>
+
+/**************************************************************************************
+ * NAMESPACE
+ **************************************************************************************/
+
+namespace impl
+{
+
+/**************************************************************************************
+ * CLASS DECLARATION
+ **************************************************************************************/
+
+class SubscriptionBase
+{
+public:
+           SubscriptionBase() { }
+  virtual ~SubscriptionBase() { }
+           SubscriptionBase(SubscriptionBase const &) = delete;
+           SubscriptionBase(SubscriptionBase &&) = delete;
+
+  SubscriptionBase & operator = (SubscriptionBase const &) = delete;
+  SubscriptionBase & operator = (SubscriptionBase &&) = delete;
+
+  virtual void onTransferReceived(CanardRxTransfer const & transfer) = 0;
+};
+
+template <typename T, typename OnReceiveCb, typename OnDestructionCb>
+class Subscription : public SubscriptionBase
+{
+public:
+  Subscription(OnReceiveCb const & on_receive_cb, OnDestructionCb const & on_destruction_cb)
+  : SubscriptionBase{}
+  , _on_receive_cb{on_receive_cb}
+  , _on_destruction_cb{on_destruction_cb}
+  { }
+  virtual ~Subscription();
+
+  virtual void onTransferReceived(CanardRxTransfer const & transfer) override;
+
+
+  CanardRxSubscription & canard_rx_subscription() { return _canard_rx_sub; }
+
+
+private:
+  CanardRxSubscription _canard_rx_sub;
+  OnReceiveCb _on_receive_cb;
+  OnDestructionCb _on_destruction_cb;
+};
+
+/**************************************************************************************
+ * NAMESPACE
+ **************************************************************************************/
+
+} /* impl */
+
+/**************************************************************************************
+ * TYPEDEF
+ **************************************************************************************/
+
+template <typename T, typename OnReceiveCb>
+using Subscription = std::shared_ptr<impl::Subscription<T, OnReceiveCb, std::function<void(void)>>>;
```

Comment:
This std::function here defeats the purpose of the latest modification to the Subscription implementation; I recommend either:

- Removing this definition completely and simply return SubscriptionBase from `Node::create_subscription()`, or
- Parameterize the last template argument as well, or
- Instead of the destroyer callback, pass a reference to the Node to the subscription object and let it do the required finalization directly.

Reply:
Thank you for your feedback :bow: 

I've opted for the final option, implemented here: 58ae355489304cdd5f05ca94a7db6e54be6f863f .